### ISSUE TITLE
Some system tray applets block auto-hide with right-click mouseover leave event 

### ISSUE THREAD

**Issue by jeremywh7** [lxqt/lxqt-panel#1266]
I encountered this recently when I added a Bluetooth dongle (blueman-applet v2.1.1).  I've seen it happen with the lxqt notification applet too, but now I can reproduce it.  It always happens with the BT applet though.

Expected Behavior: Right-click the applet to open it's menu, then click the mouse outside the panel area whether through the menu or not, and the panel should hide after the user's set delay.

Current Behavior: The above behavior results in the panel being stuck unhidden until a mouse-over or other event followed by exiting the panel again (so long as not by using an applet that triggered this issue).

Possible Solution: @palinek Seems like another edge case from: https://github.com/lxqt/lxqt/issues/853

Steps to Reproduce (for bugs)
1. Right-click Bluetooth applet
2. Move mouse outside of panel
3. Click
4. Panel does not hide

Context: I didn't see any related autohide/hide commits/issues since the 0.14.1 release.  If needed I can install the AUR GIT version for testing.

System Information:
* Distribution & Version: Arch Linux x86_64
* Kernel: Linux version 5.3.8-arch1-1 (linux@archlinux) (gcc version 9.2.0 (GCC)) SMP PREEMPT
* Qt Version: qt5-base 5.13.2-1
* liblxqt Version: 0.14.1-1
* lxqt-build-tools Version: not installed with base lxqt group in Arch
* Package version: LXQt 0.14.1
* Bluetooth packages:
 blueman 2.1.1-1
 bluez 5.52-1
 bluez-libs 5.52-1
 pulseaudio-bluetooth 13.0-2
 sbc 1.4-1

**Comment by tsujan** [comment]
I'm sure I could reproduce it with older Qt versions but can't with 5.13.1 (5.13.2 will come here soon).

Anyways, your report made me see another problem with auto-hiding; will open an issue and try to fix it later.

**Comment by tsujan** [comment]
@jeremywh7 Please test https://github.com/lxqt/lxqt-panel/pull/1267

**Comment by jeremywh7** [comment]
Had to track down my LXQt AUR build script from last year on my old PC...  :-)
Now all updated to latest git, and modded lxqt-panel-git for your PR patch (fwiw, it added 700 bytes to the pkg.tar..xz pacman install file).

The original issue is gone, but now there's another; the panel will hide as soon as I mouse-over the Bluetooth applet, instead of waiting to either leave the applet's area or click out of any applet menu/popup.

The rest of the panel and other system applets still work fine as before (like volume and removable devices).

**Comment by tsujan** [comment]
> but now there's another; the panel will hide as soon as I mouse-over the Bluetooth applet, instead of waiting to either leave the applet's area or click out of any applet menu/popup.

It's inevitable. That kind of menu doesn't belong to panel. Once it pops up, the cursor leaves the panel because the menu comes between them.

**Comment by tsujan** [comment]
> You misunderstood what I said

Here, the behavior of the right-click menu of blueman's tray icons was (and is) like other menus: no issue whatsoever. But, before the patch, I could see the issue that you reported with the *left-click* menu of lxqt-notificationd's tray icon because, contrary to its right-click menu, it doesn't belong to the panel. After the patch, when I left click it, its menu pops up and the panel hides. That's the correct behavior for menus that don't belong to panel (after applying the patch).

Now, what I understand from your last comment is that, in the case of blueman's tray icon, a mouse-over is enough to hide the panel. That definitely doesn't happen here, with blueman 2.1.1-1.

**Comment by jeremywh7** [comment]
Ah, so perhaps (~hopefully?) it's a difference with Qt 5.13.1 versus 5.13.2?  I guess we'll see.  :-) Thanks tsujan!

**Comment by tsujan** [comment]
> Ah, so perhaps (~hopefully?) it's a difference with Qt 5.13.1 versus 5.13.2?

If so, it's a serious bug in 5.13.2. Qt 5.13.2 should come here in a few days...

But we may be speaking about different things. You said, "the panel hides from just a mouse-over of the applet icon." That means you can't click the icon because the panel hides before that. Is that right?

**Comment by jeremywh7** [comment]
Correct, unless I click/right-click faster than the panel delay time (in which case the right-click menu does not prevent the panel hide like in the OP).  So simply:
1. Move the mouse cursor over toward the BT icon
2. Panel pops up as usual
2. When the cursor reaches the BT icon, the panel hides (in ~0.5sec) even without any clicking

**Comment by tsujan** [comment]
Thanks for the clear description!

Veeery strange! The hiding starts only with the "leave" event and, without a menu, there should be no "leave" event.

**Comment by tsujan** [comment]
Now I have Qt 5.13.2 but can't reproduce the second issue you mentioned. The panel does't hide before clicking and Qt 5.13.2 doesn't seem to have any bug in this regard

Actually, everything is like with Qt 5.13.1 -- as it should be because, [as the release note says](https://www.qt.io/blog/qt-5.13.2-released), "Qt 5.13.2 does not add any new functionality but provides many bug fixes and other improvements."

**Comment by Chiitoo** [comment]
With b8c33e97 [1], my panel will hide whenever I hover over anything in the tray as well.  (I'd probably expect the panel to not hide while a menu is open from any application in the tray either.)

No clicks required.

I'm on Qt 5.14 (current git master) and LXQt (current git master, minus said commit for the time being).

For fun(?), I also tried with Qt 5.12.5 and Openbox, and saw the same behaviour (using qTox as the tray test application).

1. https://github.com/lxqt/lxqt-panel/commit/b8c33e97cc7be8bddde04234240d776a100dd507

**Comment by tsujan** [comment]
@Chiitoo, Qt 5.14 isn't stable and I can't reproduce it with Qt 5.13.X.

As Qt doc says, `QEvent::Leave` is sent when "mouse leaves widget's boundaries." The systray is inside panel's boundaries. So, you're seeing an invalid leave event. Bad Qt build/installation?

**Comment by tsujan** [comment]
@Chiitoo I installed qtox minutes ago (hadn't heard of it before) and ran it. Its tray icon is normal here, like other ones: no hiding on moue-over:

![qtox](https://user-images.githubusercontent.com/981076/68631210-2872b880-04ff-11ea-8371-0ca1f877a519.png)

**Comment by Chiitoo** [comment]
> @Chiitoo, Qt 5.14 isn't stable and I can't reproduce it with Qt 5.13.X.

That's why also tried with 5.12.5.  :]

> As Qt doc says, `QEvent::Leave` is sent when "mouse leaves widget's boundaries." The systray is inside panel's boundaries. So, you're seeing an invalid leave event. Bad Qt build/installation?

The 5.12.5 test was in a VM, so there would probably have to be something wrong with Gentoo Qt as a whole.

That's good to know about qTox.  It's the first thing with a tray icon that came to mind that I could build fast-like for the VM run.  Anyblue, I'll do some more testing and see if I can find anything useful.

Thanks!

**Comment by tsujan** [comment]
> That's why also tried with 5.12.5. :]

I have an old laptop with Qt 5.12.1 (I intentionally don't upgrade its Qt) but with the latest git lxqt-panel and other LXQt components (I use it for harsh tests). I'd tested there with KWin, Openbox and Compiz-reloaded: everything was (and is) OK.

**Comment by tsujan** [comment]
@Chiitoo 

I think you might mean *the old "System tray"*.

Here, no app puts its tray icon in "System tray": All tray icons go to "Status Notifier Plugin" (which is the "new" tray), although I have added "System tray" too (in old times, Skype used it but not anymore). However, at last, I found an old java app that put its tray icon there and, yes, the panel was hidden when cursor went over it.

If so, something should be different in your system.

"System tray" is "deprecated", so to speak. I don't know (yet) why its old code may show this behavior but (1) the bug belongs to it, and (2) until it's fixed, we should either forget about "System tray" on an auto-hiding panel or re-introduce the bug that was reported here (which doesn't sound like a good idea to me): The previous code worked like a workaround for "System tray" but created the reported problem for "Status Notifier Plugin".

**Comment by Chiitoo** [comment]
Ah, yeah, that probably makes this mystery solved.

Indeed, 'tray' plugin is what I have been using.  Never knew 'statusnotifier' was for that as well, nor about the deprecation of 'tray'.

**Comment by Chiitoo** [comment]
Not sure if this should be split to a new issue, but since most of the information is here, I'll still comment on it here for now at least.

I wonder how exactly will applications decide on if they'll use the 'statusnotifier' plugin or not?  It seems that some are not able to make use of it all, Claws Mail (GTK+) being one such example.  Others, such as qTox (Qt), seem to prefer it over 'tray', if both are enabled.  Claws Mail will not have a tray icon at all if only 'statusnotifier' is enabled.

As a sidey-note, I can imagine users being a bit lost regarding the naming, as applications usually refer to a 'tray icon'.  Or maybe it's just me.  :]

Status Notifier, to me, seems like it might refer to something more specific than a general 'tray area', something something like desktop notifications perhaps.  Again though, that might just be me.

**Comment by tsujan** [comment]
> Claws Mail (GTK+) being one such example

I don't have Claws Mail but have you installed `libappindicator`? Here, even Skype's tray icon goes to Status Notifier.

The tray plugin is actually obsolete and might be removed.

**Comment by tsujan** [comment]
As for the name "Status Notifier", see https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/

**Comment by Chiitoo** [comment]
No 'libappindicator' here.  Perhaps Claws Mail should start depending on it, if it's a requirement going forward.

Thanks for the spec link!  Had not bumped into that before.  Can't say I like it after a first read at least, but I'm no designer of things.  ^^

**Comment by tsujan** [comment]
> Had not bumped into that before.

It happens years ago and all DEs -- LXQt included -- had to follow it.

**Comment by tsujan** [comment]
> Perhaps Claws Mail should start depending on it

Claws Mail seems still depending on GTK2 (at least, in Arch). If so, its code is "old", in a sense. This kind of "oldness" is forced by GTK3: GTK2 isn't developed anymore and GTK3 isn't good enough for some (or many) developers.

**Comment by Chiitoo** [comment]
Yeah, does make sense, that.  More or less obvious actually.

Thanks again!

**Comment by tsujan** [comment]
You're welcome.

I think PCMan avoided the same problem by re-writing apps in Qt. I also did so with 2 apps and, I think, many developers have a similar story -- an old story, actually.

**Comment by Chiitoo** [comment]
Yeah, I've seen some such stories unfold at distribution packaging level.

Audacious comes to mind as one example.  To quote upstream: "GTK+ 3.x support is also gone for real in this release. A -gtk3 branch still exists in GitHub, but it's not regularly updated. We barely have the resources to maintain support for two different toolkits, let alone three. Use Qt 5 if GTK+ 2.x is too "outdated" (i.e. reliable) for your liking."

I can imagine it being very annoying to the application developer.

Indeed, Audacious, when built with GTK+ 2, also seems to at least prefer 'tray'.  Seems like there's a pattern forming with regards to GTK+.

**Comment by tsujan** [comment]
I use audacious-qt5 + audacious-plugins-qt5 for years -- it's running right now, playing Internet radio :) With every new version, it gets better and I haven't had any problem with it.

As for Claws Mail, I think its develops are porting it into GTK3 but the process is slow.

> I can imagine it being very annoying to the application developer.

Developers have had 2 options: porting their programs to GTK3 or re-writhing them with C++ and Qt. Both options require lots of work but, with the first one, there's no guarantee that similar problems won't happen -- in fact, they happen again and again because GTK3 breaks codes so easily.

I don't know of an option for a unique and very complex app like Inkscape though.

**Comment by tsujan** [comment]
And last but not least ;)

Gnome3 has removed tray functionality. If that isn't done in GTK3 yet, it will sooner or later. GTK3 apps might not have tray icons at some point.

**Comment by Chiitoo** [comment]
Ah, yeah, that definitely would explain a bunch.  Have not paid attention to what happens in the GNOME lands.  Not surprising at all.

Only "issues" that I have with the Qt Audacious build, is that the global hotkeys feature is not there still (I think there is a way via mpris, but I haven't looked into it enough yet), and there are more than one taskbar entry for multiple windows in "winamp mode".  The latter really isn't an issue, since it's all minimised most of the time anyblue, and it only takes some getting used to.

Thanks yet again.  It feels good to know why these things are happening.
